### PR TITLE
[LA64_DYNAREC] Added CLC, PMOVZXBW and POPCNT opcodes

### PR DESCRIPTION
**Pull request by xiangzhai**
Hi,

Please review my patch.

Thanks,
Leslie Zhai

### CODE REVIEW

File: src/dynarec/la64/dynarec_la64_00.c [hunk 1]
```diff
@@ -2477,6 +2477,12 @@ uintptr_t dynarec64_00(dynarec_la64_t* dyn, uintptr_t addr, uintptr_t ip, int ni
                     DEFAULT;
             }
             break;
+        case 0xF8:
+            INST_NAME("CLC");
+            SETFLAGS(X_CF, SF_SUBSET);
+            SET_DFNONE();
+            BSTRINS_D(xFlags, xZR, F_CF, F_CF);
```
Comment (ksco) [inline src/dynarec/la64/dynarec_la64_00.c:2484]: This will break LBT.

File: src/dynarec/la64/dynarec_la64_f30f.c [hunk 2]
```diff
@@ -331,6 +331,51 @@ uintptr_t dynarec64_F30F(dynarec_la64_t* dyn, uintptr_t addr, uintptr_t ip, int
...
+                    VLD(v1, ed, fixedaddress);
+                } else {
+                    GETED(0); // just load and clear the upper part
+                    VINSGR2VR_D(v1, ed, 0);
+                    VINSGR2VR_D(v1, xZR, 1);
+                }
+            }
+            CLEAR_FLAGS(x2);
+            BNEZ_MARK(ed);
+            ORI(xFlags, xFlags, 1 << F_ZF);
```
Comment (ksco) [inline src/dynarec/la64/dynarec_la64_f30f.c:365]: This will break LBT.

File: src/dynarec/la64/dynarec_la64_f30f.c [hunk 3]
```diff
@@ -331,6 +331,51 @@ uintptr_t dynarec64_F30F(dynarec_la64_t* dyn, uintptr_t addr, uintptr_t ip, int
...
+            if (MODREG) {
+                if (rex.w) {
+                    VINSGR2VR_D(v1, ed, 0);
+                    VINSGR2VR_D(v1, xZR, 1);
+                } else {
+                    ZEROUP2(x4, ed); // need to clear uper part
+                    ed = x4;
+                    VINSGR2VR_D(v1, ed, 0);
+                    VINSGR2VR_D(v1, xZR, 1);
+                }
```
Comment (ksco) [inline src/dynarec/la64/dynarec_la64_f30f.c:351]: Clear v1 first with `VXOR_V` first, then you can use `VINSGR2VR_W/D` to do the job.

File: src/dynarec/la64/dynarec_la64_f30f.c [hunk 4]
```diff
@@ -331,6 +331,51 @@ uintptr_t dynarec64_F30F(dynarec_la64_t* dyn, uintptr_t addr, uintptr_t ip, int
...
+            } else {
+                if (rex.w) {
+                    SMREAD();
+                    addr = geted(dyn, addr, ninst, nextop, &ed, x1, x2, &fixedaddress, rex, NULL, 0, 0);
+                    VLD(v1, ed, fixedaddress);
```
Comment (ksco) [inline src/dynarec/la64/dynarec_la64_f30f.c:356]: This is wrong.

**Review by ksco — CHANGES_REQUESTED** [review]
The whole implementation of `POPCNT` seems to be a farfetched translation from the ARM64 backend, which I cannot approve.

**Review by ksco — CHANGES_REQUESTED** [review]
Please use `vpcnt.d` to implement this one.

I believe that copying code from RV64 or ARM64 without fully understanding it might not be the best practice and could lead to issues down the line. I'll do my best to review the code to ensure it makes sense anyway, but it would help a lot if we can avoid blind copying from other backends.

Thanks.

**Comment by xiangzhai** [comment]
> Please use vpcnt.d to implement this one.

Sorry for my stupid farfetched translation about `UADDLV` ...

File: src/dynarec/la64/dynarec_la64_f30f.c [hunk 5]
```diff
@@ -331,6 +331,48 @@ uintptr_t dynarec64_F30F(dynarec_la64_t* dyn, uintptr_t addr, uintptr_t ip, int
...
+            GETGD;
+            if (MODREG) {
+                if (rex.w) {
+                    VXOR_V(v1, v1, v1);
+                    VINSGR2VR_D(v1, ed, 0);
+                } else {
+                    ZEROUP2(x4, ed); // need to clear uper part
+                    ed = x4;
+                    VXOR_V(v1, v1, v1);
+                    VINSGR2VR_D(v1, ed, 0);
```
Comment (ksco) [inline src/dynarec/la64/dynarec_la64_f30f.c:350]: Use the _W variant here, so you don’t have to clear upper bits.

File: src/dynarec/la64/dynarec_la64_f30f.c [hunk 6]
```diff
@@ -331,6 +331,48 @@ uintptr_t dynarec64_F30F(dynarec_la64_t* dyn, uintptr_t addr, uintptr_t ip, int
...
+                }
+            } else {
+                if (rex.w) {
+                    SMREAD();
+                    addr = geted(dyn, addr, ninst, nextop, &ed, x1, x2, &fixedaddress, rex, NULL, 0, 0);
+                    FLD_D(v1, ed, fixedaddress);
+                } else {
+                    VXOR_V(v1, v1, v1);
+                    VINSGR2VR_D(v1, ed, 0);
```
Comment (ksco) [inline src/dynarec/la64/dynarec_la64_f30f.c:359]: This is still wrong, you can use FLD_S here.

File: src/dynarec/la64/dynarec_la64_f30f.c [hunk 7]
```diff
@@ -331,6 +331,48 @@ uintptr_t dynarec64_F30F(dynarec_la64_t* dyn, uintptr_t addr, uintptr_t ip, int
...
+                }
+            }
+            CLEAR_FLAGS(x2);
```
Comment (ksco) [inline src/dynarec/la64/dynarec_la64_f30f.c:362]: CLEAR FLAGS does not make sense for lbt, this should be inside the else.

File: src/dynarec/la64/dynarec_la64_f30f.c [hunk 8]
```diff
@@ -331,6 +331,48 @@ uintptr_t dynarec64_F30F(dynarec_la64_t* dyn, uintptr_t addr, uintptr_t ip, int
...
+            BNEZ_MARK(ed);
+            if (la64_lbt) {
+                ADDI_D(x3, xZR, 1 << F_ZF);
+                X64_SET_EFLAGS(x3, X_ZF);
```
Comment (ksco) [inline src/dynarec/la64/dynarec_la64_f30f.c:366]: C_ZF should be X_ALL, to set ZF but clear others.

File: src/dynarec/la64/dynarec_la64_f30f.c [hunk 9]
```diff
@@ -331,6 +331,48 @@ uintptr_t dynarec64_F30F(dynarec_la64_t* dyn, uintptr_t addr, uintptr_t ip, int
...
+                    SMREAD();
+                    addr = geted(dyn, addr, ninst, nextop, &ed, x1, x2, &fixedaddress, rex, NULL, 0, 0);
+                    FLD_D(v1, ed, fixedaddress);
+                } else {
+                    VXOR_V(v1, v1, v1);
+                    VINSGR2VR_D(v1, ed, 0);
+                }
+            }
+            CLEAR_FLAGS(x2);
+            BNEZ_MARK(ed);
```
Comment (ksco) [inline src/dynarec/la64/dynarec_la64_f30f.c:363]: What this line is for? I suppose it’s for the flags calculation, but ed here is a memory address sometimes.

**Review by ksco — CHANGES_REQUESTED** [review]

**Review by ksco — CHANGES_REQUESTED** [review]
This is the fourth round of review and the implementation is not even closer to correct. You should consider giving up this one. I can do it later, then you'll see.

**Pull request closed without merging**